raphtory/src/python/graph/node_state/lazy_node_state_latest_date_time.rs

```diff
@@ -0,0 +1,462 @@
+
+    fn __getitem__(&self, node: PyNodeRef) -> PyResult<Option<DateTime<Utc>>> {
+        let node = node.as_node_ref();
+        match self.inner.get_by_node(node) {
+            Some(v) => v.map_err(PyErr::from),
+            None => match node {
+                NodeRef::External(id) => Err(PyKeyError::new_err(format!(
+                    "Missing value for node with id {id}"
+                ))),
+                NodeRef::Internal(vid) => {
+                    let node = self.inner.graph().node(vid);
+                    match node {
+                        Some(node) => Err(PyKeyError::new_err(format!(
+                            "Missing value {}",
+                            node.repr()
+                        ))),
+                        None => Err(PyTypeError::new_err("Invalid node reference")),
+                    }
+                }
+            },
+        }
+    }
+
+    /// Iterate over items
+    ///
+    /// Returns:
+    #[doc = "     Iterator[Tuple[Node, Optional[datetime]]]: Iterator over items"]
+    fn items(&self) -> PyBorrowingIterator {
+        py_borrowing_iter_tuple_result!(
+            self.inner.clone(),
+            LazyNodeState<'static, LatestDateTime<DynamicGraph>, DynamicGraph, DynamicGraph>,
+            |inner| inner.iter().map(|(n, v)| (n.cloned(), v))
+        )
+    }
+
+    /// Iterate over values
+    ///
+    /// Returns:
+    #[doc = "     Iterator[Optional[datetime]]: Iterator over values"]
+    fn values(&self) -> PyBorrowingIterator {
+        self.__iter__()
+    }
+
+    /// Sort results by node id
+    ///
+    /// Returns:
+    #[doc = "     NodeStateResultOptionDateTime: The sorted node state"]
+    fn sorted_by_id(&self) -> NodeState<'static, LatestDateTimeOutput, DynamicGraph> {
+        self.inner.sort_by_id()
+    }
+
+    fn __repr__(&self) -> String {
+        self.inner.repr()
+    }
+
+    /// Convert results to pandas DataFrame
+    ///
+    /// The DataFrame has two columns, "node" with the node ids and "value" with
+    /// the corresponding values.
+    ///
+    /// Returns:
+    ///     DataFrame: the pandas DataFrame
+    fn to_df<'py>(&self, py: Python<'py>) -> PyResult<Bound<'py, PyAny>> {
+        let pandas = PyModule::import(py, "pandas")?;
+        let columns = PyDict::new(py);
+        columns.set_item("node", self.inner.nodes().id())?;
+        columns.set_item("value", self.values())?;
+        pandas.call_method("DataFrame", (columns,), None)
+    }
+}
+
+// impl_node_state_ord_ops
+#[pymethods]
+impl LatestDateTimeView {
+    /// Sort by value
+    ///
+    /// Arguments:
+    ///     reverse (bool): If `True`, sort in descending order, otherwise ascending. Defaults to False.
+    ///
+    /// Returns:
+    #[doc = "     NodeStateResultOptionDateTime: Sorted node state"]
+    #[pyo3(signature = (reverse = false))]
+    fn sorted(
+        &self,
+        reverse: bool,
+    ) -> NodeState<'static, Result<Option<DateTime<Utc>>, TimeError>, DynamicGraph> {
```

Comment:
this is not great, the result should probably be on the outside, otherwise it keeps failing in weird ways when you try to use the resulting NodeState instead of raising an exception up front